### PR TITLE
Auto-updates and auto-builds

### PR DESCRIPTION
**Pull request by Positive07**
# Auto-build and Auto-update
This PR fixes issues #22 and #18, providing auto-build and auto-update using electron-builder and electron-updater.

The auto-build is managed by the CI engines, Travis-CI and AppVeyor. While the updates will be fetched automatically by the app from the Github Releases.

Please note that updates are checked every time you open the app, and not constantly or periodically.

This PR requires some setup from @egoist before it can be merged: 

## Step 1: Github
* Create a [GitHub Personal Access Token](https://help.github.com/articles/creating-a-personal-access-token-for-the-command-line/), with `repo` and `admin:repo-hook` permissions

## Step 2: Travis
* Create a [Travis-CI](travis-ci.org) account
* Enable Travis-CI for devdocs-app repository
* Go to Repository Settings
* Add a [Secret Environment Variable](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings) with name `GH_TOKEN` and the access token from **Step 1** as value.
* Be sure that you clicked "ADD"

## Step 3: AppVeyor
* Create an [AppVeyor](ci.appveyor.com) account
* Add a devdocs-app as a project
* Go to Settings > Environment
* In "Environment variables" click "Add Variable"
* Create a variable with name `GH_TOKEN` and the access token from **Step 1** as value
* Go to the bottom and click "Save"

## Step 4: Draft Release
* Go to the Release section of the devdocs-app repository
* Draft a new release
* Tag the version as v0.6.0
* Save draft
* Merge this PR (After the build is complete the drafted release will have the build artifacts)
* Test the artifacts and make sure that they work fine (alternatively you can checkout [the ones I created](https://github.com/Positive07/devdocs-app/releases/tag/v0.6.0))

## Step 5: Release!
* Be happy ever after! :tada:

Just as a test you could make a minor change and push a new version, this will allow you to test the auto-update feature.

If you need help regarding the secret tokens and stuff and your prefere to discuss it privately my Discord user is Positive07@8080, or you can send me an email you can find it in my profile.

### CODE REVIEW

**Comment by Positive07** [comment]
CircleCI will fail (since there is no `circleci.yml` anymore), but this are the results from:
* Travis: [travis](https://travis-ci.org/Positive07/devdocs-app)
* AppVeyor: [appveyor](https://ci.appveyor.com/project/egoist/devdocs-app)

The results are from the `mybuilds` branch which is the same as this PR but changes the repository settings in the package.json so that the artifacts are uploaded to my fork (to which I have access through my personal token)

File: app/menu.js [hunk 1]
```diff
@@ -67,40 +58,6 @@ function createMenu(opts) {
...
-        )
-      } else {
-        dialog.showMessageBox(focusedWindow, {
-          message: 'No updates!',
-          detail: `v${pkg.version} is already the latest version.`
-        })
-      }
-    }
-  }
-
```
Comment (egoist) [inline app/menu.js]: I think we still need a menu item to check for updates manually.

Comment (Positive07) [inline app/menu.js]: I can add it yeah! But I would prefer to make it in a separate PR, I'm already working on it actually. Both that and the everyday check which existed before.

This PR concentrates on the automatic builds and the download and install of the updates. The other PR will focus on checking for updates and feedback (ie: problems with download and such)

Comment (egoist) [inline app/menu.js]: that sounds good

Comment (Positive07) [inline app/menu.js]: Great! Please take your time to review it, I can make any needed changes

**Comment by egoist** [comment]
Oh, I think we should disable updater in dev mode?

**Comment by Positive07** [comment]
Yes, that is going to happen in the next PR, error messages, periodic checks, manual check, logging.

I hope I can do all that by this weekend

**Comment by egoist** [comment]
Hmm seems periodic check is not really needed.

**Comment by Positive07** [comment]
It checks whenever you enter and there will be a way to manually check, but I would check like every 4 hours (for those that never close the app nor turn the computer off)

This is what Atom does... What do you think?

**Comment by egoist** [comment]
@Positive07 it's nice but I prefer not to use  `dialog` since it blocks the
 process.

**Comment by egoist** [comment]
Maybe just use `new Notification()` ?

**Comment by Positive07** [comment]
I don't see why a dialog is a problem, I mean what if it's blocking? If you are using app you can dismiss it, if you want to install it you accept and the app is updated, if you want to install later you dismiss it and then when you close the update is installed automatically.

So could you please explain why is dialog a problem for this?

Yes that would be great... And can be done... But that is complex in other ways

**Comment by egoist** [comment]
You need to interact with the dialog, i.e. click to dismiss or accept. It's ok to see a `dialog` when you open the app but not when you're browsing the docs.

**Comment by Positive07** [comment]
Would it really be annoying? I mean you just dismiss it... It happens every once in a while, and the notification would probably be distracting too...

Maybe there should be no popup at all... more like Chrome does it where it shows a little icon that means that next time you restart it, an update will be installed:
captura de pantalla 2017-06-17 18 39 24

PS: Can you believe that when I was writing this post I looked at that corner and an update was available? Just like I was meant to take an screenshot of it!

**Comment by egoist** [comment]
Something like that would be good 👍

**Comment by Positive07** [comment]
Where do you think it would be good to put it?

I would like the same thing to show multiple things
* Check for Updates
* Checking for Updates
* Downloading Update
* Close and Update

So it would preferably be a button with text, that changes color or something to make it easier to understand that the update has already been disabled.

Also while Checking and Downloading the button should be disabled.

**Comment by egoist** [comment]
something like this?

**Comment by Positive07** [comment]
Wow that looks cool! Can you do that injection? If you handle the renderer part I can do the main thread one. Maybe communicate them through IPC?

**Comment by egoist** [comment]
Maybe we don't need *send the "Manual Check for Updates" from the renderer to the main thread*, just add a `check for update` menu item to app menu in main process.

**Comment by Positive07** [comment]
That is a bad practice from a UX design point of view, as a user you would expect that the button to check updates will be at least in the same menu as the button where updates are reported.

And since it's not hard to talk to the renderer thread, I see no disadvantage in keeping everything in the same place.

**Pull request merged**